# Common table expressions

In [2]:
import pymysql
import sqlalchemy as alch 
from getpass import getpass
import pandas as pd
from getpass import getpass

Using the northwind database

In [3]:
password = getpass()
dbName = "northwind"
connectionData=f"mysql+pymysql://root:{password}@localhost/{dbName}"
engine = alch.create_engine(connectionData)

## 1. Write a CTE that lists the names and quantities of products with a unit price greater than $50.

```sql
-- Your code down below

```

In [22]:
query = """
WITH expensive as(
SELECT productname, unit
FROM products
WHERE price > 50)
SELECT productname, unit
FROM expensive;
"""
pd.read_sql_query(query, engine)

productname                  unit
0          Mishi Kobe Niku      18 - 500 g pkgs.
1         Carnarvon Tigers            16 kg pkg.
2   Sir Rodney's Marmalade         30 gift boxes
3  Thüringer Rostbratwurst  50 bags x 30 sausgs.
4            Côte de Blaye    12 - 75 cl bottles
5    Manjimup Dried Apples      50 - 300 g pkgs.
6     Raclette Courdavault             5 kg pkg.

#### Expected results

|ProductName                     |Unit|
|--------------------------------|----|
|Mishi Kobe Niku                 |18 - 500 g pkgs.|
|Carnarvon Tigers                |16 kg pkg.|
|Sir Rodney's Marmalade          |30 gift boxes|
|Thüringer Rostbratwurst         |50 bags x 30 sausgs.|
|Côte de Blaye                   |12 - 75 cl bottles|
|Manjimup Dried Apples           |50 - 300 g pkgs.|
|Raclette Courdavault            |5 kg pkg.|


## 2. What are the top 5 most profitable products?

```sql
-- Your code down below
```

In [26]:
query = """
WITH profitable AS (
SELECT productname, unit, SUM(price*quantity) AS total_revenue
FROM products
JOIN orderdetails ON products.productID = orderdetails.productid
GROUP BY productname, unit
ORDER BY total_revenue DESC
LIMIT 5)

SELECT productname, unit, total_revenue
FROM profitable ;
"""
pd.read_sql_query(query, engine)

productname                  unit  total_revenue
0            Côte de Blaye    12 - 75 cl bottles        63096.0
1  Thüringer Rostbratwurst  50 bags x 30 sausgs.        20832.0
2     Raclette Courdavault             5 kg pkg.        19030.0
3           Tarte au sucre               48 pies        15925.0
4        Camembert Pierrot     15 - 300 g rounds        14620.0

#### Expected results
Alice Mutton is 6th

|ProductID                       |ProductName|TotalRevenue|
|--------------------------------|-----------|------------|
|-                             |-|-       |
|-                           |-|-       |
|-                            |-|-       |
|-                             |-|-       |
|-                              |-|-       |
|17                              |Alice Mutton|12909       |

## 3. Write a CTE that lists the top 5 categories by the number of products they have.

```sql
-- Your code down below
```

In [27]:
query = """
WITH categories as(
SELECT CategoryName, COUNT(products.categoryID) AS ProductCount
FROM categories
JOIN products ON categories.categoryID = products.categoryID
GROUP BY CategoryName
ORDER BY ProductCount DESC
LIMIT 5)

SELECT CategoryName, ProductCount
FROM categories;
"""
pd.read_sql_query(query, engine)

CategoryName  ProductCount
0     Confections            13
1       Beverages            12
2      Condiments            12
3         Seafood            12
4  Dairy Products            10

#### Expected result
|CategoryName                    |ProductCount|
|--------------------------------|------------|
|Confections                     |13          |
|Beverages                       |12          |
|Condiments                      |12          |
|Seafood                         |12          |
|Dairy Products                  |10          |


## 4. Write a CTE that shows the average order quantity for each product category.

```sql
-- Your code down below

```

In [28]:
query = """
WITH avgquant as(
SELECT CategoryName, AVG(orderdetails.quantity) AS AvgOrderQuantity
FROM categories
JOIN products ON categories.categoryID = products.categoryID
JOIN OrderDetails ON products.productID = OrderDetails.productID
GROUP BY CategoryName
ORDER BY AvgOrderQuantity DESC)

SELECT CategoryName, AvgOrderQuantity
FROM avgquant;
"""
pd.read_sql_query(query, engine)

CategoryName  AvgOrderQuantity
0      Condiments           28.2245
1  Dairy Products           26.0100
2    Meat/Poultry           25.7600
3     Confections           25.1190
4       Beverages           24.6129
5  Grains/Cereals           21.7143
6         Produce           21.6667
7         Seafood           21.5672

|CategoryName                    |AvgOrderQuantity|
|--------------------------------|----------------|
|Beverages                       |24.6129         |
|Condiments                      |28.2245         |
|Confections                     |25.1190         |
|Dairy Products                  |26.0100         |
|Grains/Cereals                  |21.7143         |
|Meat/Poultry                    |25.7600         |
|Produce                         |21.6667         |
|Seafood                         |21.5672         |


# 5. Create a CTE to calculate the average order amount for each customer.

```sql
-- Your code down below


```

In [101]:
query = """
WITH orderamount as(
SELECT customers.CustomerID, CustomerNAme, AVG(quantity*price) AS AvgOrderAmount
FROM customers
JOIN orders ON customers.customerID = orders.customerID
JOIN OrderDetails ON orders.orderID = OrderDetails.orderID
JOIN Products ON OrderDetails.productID = Products.productID
GROUP BY orders.OrderID, CustomerID
ORDER BY AvgOrderAmount DESC)

SELECT CustomerID, CustomerName, AvgOrderAmount
FROM orderamount;
"""
pd.read_sql_query(query, engine)

CustomerID                        CustomerName  AvgOrderAmount
0            59                    Piccolo und mehr         6726.00
1            51                      Mère Paillarde         4802.00
2            62                       Queen Cozinha         3845.50
3            73                       Simons bistro         3536.75
4            55              Old World Delicatessen         2200.00
..          ...                                 ...             ...
191          13          Centro comercial Moctezuma           63.00
192          27                      Franchi S.p.A.           62.00
193           2  Ana Trujillo Emparedados y helados           55.50
194          66                  Reggiani Caseifici           54.50
195          69                    Romero y tomillo           37.00

[196 rows x 3 columns]

|CustomerID                      |CustomerName|AvgOrderAmount|
|--------------------------------|------------|--------------|
|59                              |Piccolo und mehr|4014.2500     |
|73                              |Simons bistro|2444.3333     |
|62                              |Queen Cozinha|1991.6667     |
|51                              |Mère Paillarde|1673.8571     |
|71                              |Save-a-lot Markets|1407.2500     |
|76                              |Suprêmes délices|1345.8333     |
|81                              |Tradição Hipermercados|1315.6667     |
|7                               |Blondel père et fils|1174.4615     |
|89                              |White Clover Markets|1112.5000     |
|55                              |Old World Delicatessen|1079.5000     |
|20                              |Ernst Handel|1018.0000     |
|19                              |Eastern Connection|1004.8000     |
|68                              |Richter Supermarkt|976.6667      |
|72                              |Seven Seas Imports|934.0000      |
|63                              |QUICK-Stop  |908.5500      |
|25                              |Frankenversand|895.0667      |
|75                              |Split Rail Beer & Ale|854.7692      |
|65                              |Rattlesnake Canyon Grocery|838.2727      |
|52                              |Morgenstern Gesundkost|754.0000      |
|9                               |Bon app''   |750.4286      |
|37                              |Hungry Owl All-Night Grocers|733.5714      |
|35                              |HILARIÓN-Abastos|722.8333      |
|33                              |GROSELLA-Restaurante|690.0000      |
|34                              |Hanari Carnes|681.0000      |
|15                              |Comércio Mineiro|677.5000      |
|23                              |Folies gourmandes|672.3333      |
|10                              |Bottom-Dollar Marketse|664.0000      |
|31                              |Gourmet Lanchonetes|641.0000      |
|60                              |Princesa Isabel Vinhoss|628.5000      |
|5                               |Berglunds snabbköp|601.5556      |
|46                              |LILA-Supermercado|564.7692      |
|44                              |Lehmanns Marktstand|545.7500      |
|30                              |Godos Cocina Típica|515.0000      |
|3                               |Antonio Moreno Taquería|504.0000      |
|88                              |Wellington Importadora|503.8333      |
|47                              |LINO-Delicateses|500.0000      |
|56                              |Ottilies Käseladen|500.0000      |
|87                              |Wartian Herkku|492.5833      |
|24                              |Folk och fä HB|480.2222      |
|49                              |Magazzini Alimentari Riuniti|464.2857      |
|86                              |Die Wandernde Kuh|441.9091      |
|41                              |La maison d''Asie|437.8182      |
|8                               |Bólido Comidas preparadas|416.0000      |
|83                              |Vaffeljernet|411.0000      |
|14                              |Chop-suey Chinese|374.8333      |
|80                              |Tortuga Restaurante|357.4167      |


# 6. Sales Analysis with CTEs

Assume we have the Northwind database which contains tables like Orders, OrderDetails, and Products. Create a CTE that calculates the total sales for each product in the year 1997.

```sql
-- Your code down below

 ```

In [76]:
query = """
WITH totalsales as(
SELECT ProductName, SUM(quantity) AS TotalSales
FROM Products
JOIN OrderDetails ON products.productID = OrderDetails.productID
JOIN orders ON OrderDetails.OrderID = orders.OrderID
WHERE OrderDate BETWEEN '1997-01-01' AND '1997-12-31'
GROUP BY ProductName
ORDER BY TotalSales DESC)

SELECT ProductName, TotalSales
FROM totalsales;
"""
pd.read_sql_query(query, engine)

ProductName  TotalSales
0             Gnocchi di nonna Alice       173.0
1                          Tourtière       126.0
2                            Geitost       119.0
3                              Chang       115.0
4               Raclette Courdavault       115.0
5                     Sirop d'érable       106.0
6                       Vegie-spread       100.0
7                      Côte de Blaye        99.0
8                       Alice Mutton        97.0
9                     Steeleye Stout        95.0
10               Sir Rodney's Scones        92.0
11                           Pavlova        86.0
12                     Zaanse koeken        85.0
13                       Fløtemysost        75.0
14                    Tarte au sucre        75.0
15               Gumbär Gummibärchen        72.0
16                    Queso Cabrales        72.0
17                         Chocolade        70.0
18                              Tofu        67.0
19               Scottish Longbreads        66.0
20         Louisiana Hot Spiced Okra        60.0
21                          Maxilaku        60.0
22                 Camembert Pierrot        60.0
23                          Tunnbröd        60.0
24        Teatime Chocolate Biscuits        57.0
25                 Rössle Sauerkraut        54.0
26                      Lakkalikööri        52.0
27                  Boston Crab Meat        52.0
28                Schoggi Schokolade        50.0
29                     Aniseed Syrup        50.0
30           Thüringer Rostbratwurst        45.0
31   Original Frankfurter grüne Soße        45.0
32                     Longlife Tofu        45.0
33         Wimmers gute Semmelknödel        43.0
34                      Gula Malacca        40.0
35                     Perth Pasties        40.0
36                  Valkoinen suklaa        40.0
37     Singaporean Hokkien Fried Mee        40.0
38                  Gudbrandsdalsost        35.0
39                    Ravioli Angelo        35.0
40                             Chais        34.0
41                         Spegesild        27.0
42   Jack's New England Clam Chowder        25.0
43              Rhönbräu Klosterbier        24.0
44                   Mishi Kobe Niku        20.0
45                        Röd Kaviar        20.0
46  Louisiana Fiery Hot Pepper Sauce        20.0
47                     Sasquatch Ale        20.0
48             Nord-Ost Matjeshering        18.0
49         Queso Manchego La Pastora        15.0
50                 Gorgonzola Telino        14.0
51               Gustaf's Knäckebröd        12.0
52                      Pâté chinois        10.0
53            Mozzarella di Giovanni        10.0
54                        Gravad lax        10.0
55                     Outback Lager         8.0
56                       Inlagd Sill         5.0
57                             Konbu         2.0

#### Expected result

|ProductName                     |TotalSales|
|--------------------------------|----------|
|Gnocchi di nonna Alice          |173       |
|Tourtière                       |126       |
|Geitost                         |119       |
|Chang                           |115       |
|Raclette Courdavault            |115       |
|Sirop d'érable                  |106       |
|Vegie-spread                    |100       |
|Côte de Blaye                   |99        |
|Alice Mutton                    |97        |
|Steeleye Stout                  |95        |
|Sir Rodney's Scones             |92        |
|Pavlova                         |86        |
|Zaanse koeken                   |85        |
|Fløtemysost                     |75        |
|Tarte au sucre                  |75        |
